### ISSUE TITLE
Onboarding: Unable to navigate back to wizard, responses not saved

### ISSUE THREAD

**Issue by stephhjacq** [woocommerce/woocommerce-admin#3067]
**Describe the bug**
I tested the new onboarding starting with the Profile Wizard. 

During the wizard, I selected that I would like to use Subscriptions (as well as physical products and downloads), and was expecting some sort of purchase flow as this was priced at $199, but this selection seemed to do nothing.

![https://cld.wthms.co/qd7ceg](https://cld.wthms.co/qd7ceg+)
Full Image: https://cld.wthms.co/qd7ceg+

I selected Mailchimp, but nothing was installed.

![https://cld.wthms.co/D1Qzo7](https://cld.wthms.co/D1Qzo7+)
Full Image: https://cld.wthms.co/D1Qzo7+

![https://cld.wthms.co/EzXjwE](https://cld.wthms.co/EzXjwE+)
Full Image: https://cld.wthms.co/EzXjwE+

After completing the wizard I was sent to the task list, where I did the first task, `Create home page`. Once that was published, I could not return to the wizard.

Unclear where to go from here:

![https://cld.wthms.co/plxctT](https://cld.wthms.co/plxctT+)
Full Image: https://cld.wthms.co/plxctT+

Going to WooCommerce > Dashboard takes me back to the beginning of the profile wizard, which has not recorded any of my responses:

![https://cld.wthms.co/kT8zrT](https://cld.wthms.co/kT8zrT+)
Full Image: https://cld.wthms.co/kT8zrT+

I tried using the other flow, WooCommerce > Orders > Help > Setup Wizard, however the Orders screen has no help menu.

![https://cld.wthms.co/Q2CfWP](https://cld.wthms.co/Q2CfWP+)
Full Image: https://cld.wthms.co/Q2CfWP+

I found the Help menu under Products, and was able to find the task list by disabling the profile wizard.

I was able to continue with the task list, however I ran across the following error when connecting Stripe, you must be an admin:

![https://cld.wthms.co/T5zNZg](https://cld.wthms.co/T5zNZg+)
Full Image: https://cld.wthms.co/T5zNZg+

I have two profiles on Stripe.com under this account - one for WooCommerce/Bubblestorm and one for my test site. Not sure if that makes me an edge-case, maybe selecting the profile is not required for the average user. 

After exiting the payment setup, I re-enabled the Profile Wizard. I re-entered all my information, and it seemed to complete this time- now the task list is shown when I go to WooCommerce > Dashboard, and the Wizard is disabled under the help menu. 

**Expected behavior**
I selected that I would like to use Subscriptions (as well as physical products and downloads), and was expecting some sort of purchase flow as this was priced at $199, but this selection seemed to do nothing.

I also selected Mailchimp and the notice indicated the plugin would be installed, it was not.

After completing a task in the task list, I expected a way to navigate back to the list to continue. 

**Desktop (please complete the following information):**

* OS: 10.14.6
* Browser Chrome
* Version 77.0.3865.120

**Comment by joshuatf** [comment]
Thanks so much for testing, @stephhjacq!

>During the wizard, I selected that I would like to use Subscriptions (as well as physical products and downloads), and was expecting some sort of purchase flow as this was priced at $199, but this selection seemed to do nothing.

This is one of the last big unfinished pieces for the profiler.  This will redirect to WCCOM and I think there's nearly a working solution there currently, but last I checked we might still need a redirect parameter back to the store site. /cc @mattyza 

https://github.com/woocommerce/woocommerce-admin/issues/2477

> I selected Mailchimp, but nothing was installed.

I'm not sure if we plan to install these extensions as part of the above process, but /cc @justinshreve to double check.  We could also go ahead and install these async after the "Business Details" step.

>After completing the wizard I was sent to the task list, where I did the first task, Create home page. Once that was published, I could not return to the wizard.

Ready to merge update for that [here](https://github.com/woocommerce/woocommerce-admin/pull/3061).

>Going to WooCommerce > Dashboard takes me back to the beginning of the profile wizard, which has not recorded any of my responses:

This one is a bit troubling.  Did you toggle the Onboarding Profiler on at any point between the task list?  Or were you working through the task list and somehow found yourself with this profiler screen once again re-enabled.  Would you mind writing out a full list of steps to recreate the issue?

>I tried using the other flow, WooCommerce > Orders > Help > Setup Wizard, however the Orders screen has no help menu.

I dug through WC a bit and found that the `#screen-meta-links` tab is hidden on empty state (when no orders are present).  This is an interesting choice; I think it shouldn't be too much of an issue for us since the toggling of the profiler/tasks here is mostly for debugging, but maybe someone more knowledgeable than myself can chime in as to why the "Help" tab is hidden.

>I was able to continue with the task list, however I ran across the following error when connecting Stripe, you must be an admin:

I'm not sure there's a whole lot we can do on the Stripe side if you don't have the correct privileges on the account.  I do notice a "Switch user" in the upper right of that screenshot.  Do you feel this would be confusing to users?

My thoughts are that most users (if they have multiple Stripe accounts) may be familiar with switching, but open to other opinions here.

**Comment by mattyza** [comment]
> > During the wizard, I selected that I would like to use Subscriptions (as well as physical products and downloads), and was expecting some sort of purchase flow as this was priced at $199, but this selection seemed to do nothing.
> 
> This is one of the last big unfinished pieces for the profiler. This will redirect to WCCOM and I think there's nearly a working solution there currently, but last I checked we might still need a redirect parameter back to the store site. /cc @mattyza

I believe this can now use the in-app purchase flow's parameters for creating a "back" link.

https://github.com/woocommerce/woocommerce/blob/master/includes/admin/class-wc-admin-addons.php#L490

```
'wccom-site'          => site_url(),
'wccom-back'          => esc_url( $back_admin_path ),
'wccom-woo-version'   => WC_VERSION,
'wccom-connect-nonce' => wp_create_nonce( 'connect' ),
```

@joshuatf

**Comment by stephhjacq** [comment]
Hey @joshuatf, you're welcome, happy to help 👍  

>>I selected Mailchimp, but nothing was installed.

>I'm not sure if we plan to install these extensions as part of the above process, but /cc @justinshreve to double check. We could also go ahead and install these async after the "Business Details" step.

Sure, just noted that the text says it will be installed, so that would need to be removed.

>>Going to WooCommerce > Dashboard takes me back to the beginning of the profile wizard, which has not recorded any of my responses:

>This one is a bit troubling. Did you toggle the Onboarding Profiler on at any point between the task list? Or were you working through the task list and somehow found yourself with this profiler screen once again re-enabled. Would you mind writing out a full list of steps to recreate the issue?

I did not toggle. Here's my steps, Attempt 1:
1. Worked through the wizard questions
2. Next screen after completing wizard is the task list
3. Begin first task - set up home page. I add a little text, hit Publish.
4. After publishing that page, there is nothing to indicate what I should do next or how to get back to the list
5. Trying to find the list, I click WooCommerce > Dashboard, and see the first step of the wizard again, with empty fields.

Attempt 2:
1. Worked through the wizard questions (again)
2. Next screen after completing wizard is the task list
3. This shows the first few tasks as completed already (expected, I did these in Attempt 1)
4. Now when I go back to WooCommerce > Dashboard I see the task list (instead of the wizard), and when I go to the help screen I see the wizard is disabled (seems it is now registering as completed)

>>I was able to continue with the task list, however I ran across the following error when connecting Stripe, you must be an admin:

>I'm not sure there's a whole lot we can do on the Stripe side if you don't have the correct privileges on the account. I do notice a "Switch user" in the upper right of that screenshot. Do you feel this would be confusing to users?

>My thoughts are that most users (if they have multiple Stripe accounts) may be familiar with switching, but open to other opinions here.

The issue is that I have only a single user account (my a8c address), with access to two Stripe accounts. I would not expect this to be an issue for most of our users (hopefully?) but wanted to note it just in case.

**Comment by justinshreve** [comment]
Re Facebook & MailChimp @joshuatf :

> I'm not sure if we plan to install these extensions as part of the above process, but /cc @justinshreve to double check. We could also go ahead and install these async after the "Business Details" step.

I was thinking that we would also pass these to WooCommerce.com to be installed, especially since for the Facebook plugin, support is offered for that year:

* https://woocommerce.com/products/facebook/
* https://woocommerce.com/products/mailchimp-for-woocommerce/

**Comment by joshuatf** [comment]
>If they don't select any extensions to purchase / themes -- just async installing them would be nicer for the user then having to figure out another login/sign-up flow right away.

@justinshreve Ya, I think this might be the best way.  Ideally we could install through WCCOM, but if that portion gets skipped in one way or another, it would be nice to still have those plugins installed to avoid confusion.

**Comment by joshuatf** [comment]
>Worked through the wizard questions
Next screen after completing wizard is the task list
Begin first task - set up home page. I add a little text, hit Publish.
After publishing that page, there is nothing to indicate what I should do next or how to get back to the list
Trying to find the list, I click WooCommerce > Dashboard, and see the first step of the wizard again, with empty fields.

@justinshreve @psealock I wasn't able to recreate the issue following these steps on a Jurassic Ninja site.  Can you see if either of you are able to?

@stephhjacq Could I bother you to run through the process one more time when you have time, keeping your browser console open and let me know if any errors or notices pop up while you're going through the onboarding profiler?

@justinshreve Should we overwrite the existing release with an updated one to eliminate some of the bugs we've handled or create a new one?

**Comment by justinshreve** [comment]
> I wasn't able to recreate the issue following these steps on a Jurassic Ninja site. Can you see if either of you are able to?

@joshuatf I just spun up .a new Jurassic Ninja site and was not able to recreate the profile wizard issue using the currently released zip file. When I navigated back to the dashboard, the task list displayed correctly.

**Comment by justinshreve** [comment]
@joshuatf I _think_ I know what's happening here and I believe it's related to https://github.com/woocommerce/woocommerce-admin/issues/2959.

We set the chosen theme in the profile object in one request, and then we set completed in a separate request.

https://github.com/woocommerce/woocommerce-admin/blob/6f774d2eb7155f42650d33632c14872e846c45f9/src/API/OnboardingProfile.php#L128-L143 has to pull from the current option and merge with the passed updates. It seems to me like maybe these requests complete at different times so the complete status can get trampled.

I ran into this while implementing the cart logic, so I ended up moving the logic around just a bit for this specific case that I believe should prevent this from happening (though we will obviously still see bugs in other places that use this 'pattern').

https://github.com/woocommerce/woocommerce-admin/pull/3153/files#diff-6fa47cc483d08710475e62cd351597a1R53-R74

I think we can probably close this out since we are tracking the async issue elsewhere, and my PR should fix this specific issue.

**Comment by stephhjacq** [comment]
@joshuatf sorry for the delay, I've just tested this again. I used a different test site this time.

![https://cld.wthms.co/8KKTvm](https://cld.wthms.co/8KKTvm+)
Full Image: https://cld.wthms.co/8KKTvm+

Is it intended to make Jetpack compulsory? Many stores do not want to install this. Additionally, we allow CBD vendors to sell on WooCommerce but they cannot connect to Jetpack. Can we make this optional?

![https://cld.wthms.co/CpjIGf](https://cld.wthms.co/CpjIGf+)
Full Image: https://cld.wthms.co/CpjIGf+

I saw a message in the console `Uncaught in promise user is unauthorized` or similar, but I didn't get to capture it before it switched to the above error.

Clicking 'Return to site' took me to WordPress.com instead of my site.

Navigating back to my site and starting the Wizard again, I get this:

![https://cld.wthms.co/a8ZIDo](https://cld.wthms.co/a8ZIDo+)
Full Image: https://cld.wthms.co/a8ZIDo+

Some 404s from Mailchimp here

![https://cld.wthms.co/MJyXR7](https://cld.wthms.co/MJyXR7+)
Full Image: https://cld.wthms.co/MJyXR7+

Errors after publishing the homepage

![https://cld.wthms.co/VfRNHs](https://cld.wthms.co/VfRNHs+)
Full Image: https://cld.wthms.co/VfRNHs+

And now when I go to WooCommerce > Dashboard, it's forgotten my details. No errors though:

![https://cld.wthms.co/89ewLF](https://cld.wthms.co/89ewLF+)
Full Image: https://cld.wthms.co/89ewLF+

After running through the wizard a second time, I'm taken to the task list. 

The issue with the wizard forgetting the details may be related to the home page publication task - the second time I run the wizard I do not do this (as the page is already published), and I do not see the issue.

Let me know if I can provide further info.

**Comment by justinshreve** [comment]
I believe all of the reports in here have been handled, so I’m closing this issue.

#3154 and #3165 should specifically help with navigating between the wizard and the responses being saved.

Walking back through the other issues reported here:

The Gutenberg home page notice is addressed by #3061 and #3206.

MailChimp & Facebook are installed as of #3083. Images fixed in #3062.

The purchase flow is wired up in #3121 and #3173.

The condition outlined here https://github.com/woocommerce/woocommerce-admin/issues/3067#issuecomment-548545478 has been fixed.
